cpp-ch/local-engine/Operator/ExpandTransform.cpp

```diff
@@ -68,43 +67,36 @@ ExpandTransform::Status ExpandTransform::prepare()
 void ExpandTransform::work()
 {
     assert(expanded_chunks.empty());
-    size_t agg_cols_size = aggregating_expressions_columns.size();
-    for (int set_id = 0; static_cast<size_t>(set_id) < grouping_sets.size(); ++set_id)
+    for (const auto & project_exprs: project_set_exprs)
     {
-        const auto & sets = grouping_sets[set_id];
-        DB::Columns cols;
         const auto & original_cols = input_chunk.getColumns();
-        for (size_t i = 0; i < original_cols.size(); ++i)
+        size_t rows = input_chunk.getNumRows();
+        DB::Columns cols;
+        for (const auto & project_expr : project_exprs)
         {
-            const auto & original_col = original_cols[i];
-            size_t rows = original_col->size();
-            if (i < agg_cols_size)
-            {
-                cols.push_back(original_col);
-                continue;
-            }
-            // the output columns should all be nullable.
-            if (!sets.contains(i))
+            if (project_expr.kind == EXPAND_FIELD_KIND_SELECTION)
             {
-                auto null_map = DB::ColumnUInt8::create(rows, 1);
-                auto col = DB::ColumnNullable::create(original_col, std::move(null_map));
-                cols.push_back(std::move(col));
+                cols.push_back(original_cols[project_expr.field.get<Int32>()]);
```

Comment:
Do you have tested the case grouping on a key which is non-nullable ?

Reply:
case like that 
```SQL
create table test(
  a Int32,
  b Nullable(Int32)
);
```

``` SQL
select a, b, count(1) from test group by a,b with rollup
```

In vallina spark, I guess the `a` in the result of `expand` should be nullable